File: Army(java)/Basic/SpellCaster.java
package Basic;

import EnumID.*;
import State.*;
import Magic.*;

public abstract class SpellCaster extends Unit {
    protected MagicPower magic;
    protected SpellBook book;

    public SpellCaster(String name, int hp, int mana, int dmg){
        super(name, hp, dmg);
        this.magic = new MagicPower(mana);
        this.book = new SpellBook();
    }

    public abstract MagicianID getMagicianID();

    public MagicPower getMagicPower() {
        return magic;
    }

    public SpellBook getSpellBook() {
        return book;
    }

    public void magicAttack(Unit enemy, String spell) throws UnitIsDeadException, MagicManaOffException {
        if ( spell == "Freeze" || spell == "HellFire" || spell == "Lightning" || spell == "AcidMist" ) {           
            book.getSpellBook().get(spell).mAttack(enemy, this);
        }
    }

    public void healHealth(Unit other, String spell) throws UnitIsDeadException, MagicManaOffException {
        if ( spell == "Heal" ) {
            book.getSpellBook().get(spell).healHealth(other, this);
        } else if ( spell == "Human" ) {
            book.getSpellBook().get(spell).backToHuman(other, this);
        } else {
            return;
        }
    }

    public void healMagic(SpellCaster other) throws UnitIsDeadException, MagicManaOffException {
        book.getSpellBook().get("ReturnMagic").healMagic(other, this);
    }

    public void printMagician() {
        this.printUnit();
        getMagicPower().printPower();
    }
}

File: Army(c++)/Basic/SpellCaster.cpp
#include "SpellCaster.hpp"
#include "../Magic/SpellBook.hpp"

SpellCaster::SpellCaster(const std::string& name,  int hp, int mana, int dmg, State* state) 
    : Unit(name, hp, dmg, state), magic(new MagicPower(mana))
    , book(new SpellBook()) {}

SpellCaster::~SpellCaster(){
    delete this->magic;
    delete this->book;
}

MagicPower& SpellCaster::getMagicPower() const {
    return *(this->magic);
}

SpellBook& SpellCaster::getSpellBook() const {
    return *(this->book);
}

void SpellCaster::magicAttack(Unit* enemy, SpellID spell) {
    if ( spell == FREEZE || spell == HELLFIRE || spell == LIGHTNING || spell == ACIDMIST ) { 
        this->book->getSpellBook().find(spell)->second->mAttack(enemy, this);
    }
}

void SpellCaster::healHealth(Unit* other, SpellID spell) {
    if ( spell == HEAL ) {
        this->book->getSpellBook().find(spell)->second->healHealth(other, this);
    } else if ( spell == HUMAN ) {
        this->book->getSpellBook().find(spell)->second->backToHuman(other, this);
    } else {
        return;
    }
}

void SpellCaster::healMagic(SpellCaster* other) {
    this->book->getSpellBook().find(RETURN_MAGIC)->second->healMagic(other, this);
}

std::ostream& operator<<(std::ostream& out, const SpellCaster& spellCaster) {
    out << spellCaster.getName() << " - " << typeid(spellCaster.getState()).name() << std::endl;
    out << spellCaster.getMagicPower() << std::endl;
    out << spellCaster.getHealth() << std::endl;
    
    return out;
}
